atArgumentParser.go

```diff
@@ -2,102 +2,171 @@ package vmcommon
 
 import (
 	"encoding/hex"
-	"strings"
 )
 
-type atArgumentParser struct {
+// AtArgumentParser is a parser that splits arguments by @ character
+// [NotConcurrentSafe]
+type AtArgumentParser struct {
+	// First argument is a string (function name or hex-encoded bytecode), the rest are raw bytes
 	arguments [][]byte
-	code      []byte
 }
 
-const atSep = "@"
-const atSepChar = '@'
+const atSeparator = "@"
+const atSeparatorChar = '@'
+const startIndexOfConstructorArguments = 3
+const startIndexOfFunctionArguments = 1
+const minNumDeployArguments = 3
+const minNumCallArguments = 1
+const indexOfCode = 0
+const indexOfVMType = 1
+const indexOfCodeMetadata = 2
+const indexOfFunction = indexOfCode
+
+// NewAtArgumentParser creates a new parser
+func NewAtArgumentParser() *AtArgumentParser {
+	parser := &AtArgumentParser{}
+	parser.clearArguments()
+	return parser
+}
 
-// NewAtArgumentParser creates a new argument parser implementation that splits arguments by @ character
-func NewAtArgumentParser() *atArgumentParser {
-	return &atArgumentParser{}
+func (parser *AtArgumentParser) clearArguments() {
+	parser.arguments = make([][]byte, 0)
 }
 
-// ParseData creates the code and the arguments from the input data
-// format: code@arg1@arg2@arg3...
-// Until the first @ all the bytes are for the code / function
-// after that every argument start with an @
-func (at *atArgumentParser) ParseData(data string) error {
-	splitString := strings.Split(data, atSep)
-	if len(splitString) == 0 || len(splitString[0]) == 0 {
-		return ErrStringSplitFailed
+// ParseData parses strings of the following formats:
+// contract deploy: codeHex@vmTypeHex@codeMetadataHex@argFooHex@argBarHex...
+// contract call: functionRaw@argFooHex@argBarHex...
+func (parser *AtArgumentParser) ParseData(data string) error {
+	parser.clearArguments()
+
+	tokens := tokenize(data)
+	err := requireAnyTokens(tokens)
```

Comment:
Consider merging `tokenize` and `requireAnyTokens` into a single function `tokenize`, which returns `tokens, err`.

Reply:
Merged.